Failed Authentication

annym opened the pull request:
## Overview

- Corfu Server log statement to report the target IP that failed to authenticate.
- Client exponential back off for reconnects upon authentication failure.

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java:
```diff
@@ -462,10 +460,29 @@ protected void initChannel(@Nonnull Channel ch) throws Exception {
+                    sslHandler.handshakeFuture().addListener(channelFuture -> {
+                        if (channelFuture.cause() != null && channelFuture.cause().getClass()
+                                .equals(SSLHandshakeException.class)) {
+                            String peerHost;
```
Can't you just use 
`ch.remoteAddress()`
instead of casting?

Example:
https://github.com/CorfuDB/CorfuDB/blob/master/runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java#L153

annym replied:
Done. It also prints the port info but no harm in this.

Maithem replied:
On re-connection the client will bind to a new port, that's how I was able to tell that the client was spinning on reconnecting (the ports kept changing on every request)

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java:
```diff
@@ -462,10 +460,29 @@ protected void initChannel(@Nonnull Channel ch) throws Exception {
+                            } catch (Exception e) {
+                                peerHost = "Not determined.";
+                            }
+                            log.error("CorfuServer authentication failed for target IP: {}", peerHost);
```
Can you revise the wording? The clients try to authenticate with the server so is the ip really a target?
Consider:
`Client authentication failed (remote address {address})`

Maithem commented on runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java:
```diff
@@ -191,6 +209,10 @@ public NettyClientRouter(@Nonnull NodeLocator node,
 
+        sslAuthenticationFailure = new AtomicBoolean();
+        retryPolicy = new ExponentialBackoffRetry(null);
+        retryPolicy.setMaxRetryThreshold(maxRetryThreshold);
```
Can you make the ExponentialBackoffRetry object immutable and pass the the threshold in the constructor ?

Maithem commented on runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java:
```diff
@@ -330,6 +352,12 @@ private void addReconnectionOnCloseFuture(@Nonnull Channel channel,
+                    retryPolicy.nextWait();
+                } else {
+                    // reset exponential time retry policy, to start from beginning on next failure.
+                    retryPolicy.reset();
```
This seems a bit strange if you have an object that encapsulates the exponential backoff logic why do you need to invoke reset externally?  Shouldn't retryPolicy.nextWait reset itself internally ?

annym replied:
This reset happens in between authentication failures and successes to reset the timer calculation from the beginning and exponentially backoff from the start. Otherwise, in an scenario failure, success, failure, this last would never restart it would continue from the rate of the last backoff.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java:
```diff
@@ -330,6 +352,12 @@ private void addReconnectionOnCloseFuture(@Nonnull Channel channel,
             if (!shutdown) {
                 log.info("addReconnectionOnCloseFuture[{}]: reconnecting", node);
+                if (sslAuthenticationFailure.get()) {
+                    retryPolicy.nextWait();
```
Can we make nextWait return the next value, so that we can invoke the sleep logic directly? 
That way we can add logging with the right context.

annym replied:
Done

Maithem commented on runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java:
```diff
@@ -629,6 +662,16 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
+                // Event fired once the SSL handshake is complete, which may have succeeded or not.
+                // If SSL handshake failed, activate flag to exponentially backoff reconnection time.
+                SslHandshakeCompletionEvent sslEvt = (SslHandshakeCompletionEvent) evt;
+                if (sslEvt.cause() instanceof SSLHandshakeException) {
```
I don't think this fix should be specific to ssl handshake failures. We should should solve the general problem (i.e. clients spinning on errors). Any exception can cause the channel to be closed, for example:
1. client/server handshake 
2. authentication failures
3. encode/decode failures
4. other errors

In all of those cases, we should have an exponential backoff.

apavan replied:
Good points! Is your suggestion to handle all these as part of this PR or create issues that could be added to our backlog.

annym replied:
From the logs I could observe these exceptions are wrapped by the SSLHandshakeException.

Maithem replied:
I don't think these cases are handled differently. We just need to apply the exponential backoff to the reconnect logic.

StrikingThirteen replied:
> From the logs I could observe these exceptions are wrapped by the SSLHandshakeException.

Why endcode/decode failures are wrapped by SshHandshakeException?

Maithem replied:
When an exception happens during encode/decode we close the channel and disconnect, then subsequent requests will try to open a new channel thus triggering the reconnect logic.

StrikingThirteen commented on runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java:
```diff
@@ -128,6 +136,14 @@
+     * Indicates if SSL authentication has failed or not, in order to apply a backoff mechanism for
+     * exponential time reconnection.
+     */
+    private AtomicBoolean sslAuthenticationFailure;
```
isSslAuthenticationFailed is more expressive (and then optionally remove the variable comment)

StrikingThirteen commented on runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java:
```diff
@@ -606,6 +636,16 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
+                // Event fired once the SSL handshake is complete, which may have succeeded or not.
+                // If SSL handshake failed, activate flag to exponentially backoff reconnection time.
+                SslHandshakeCompletionEvent sslEvt = (SslHandshakeCompletionEvent) evt;
+                if (sslEvt.cause() instanceof SSLHandshakeException) {
```
There are several levels of nested if else in the code. Can you refactor?

StrikingThirteen commented on runtime/src/main/java/org/corfudb/util/retry/ExponentialBackoffRetry.java:
```diff
@@ -83,4 +88,8 @@ public void nextWait() {
         Sleep.MILLISECONDS.sleepUninterruptibly(sleepTime);
     }
 
+    public void reset() {
```
Why not reseting the retryCounter in the reset method?

annym replied:
Good catch. Done.

StrikingThirteen commented on runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java:
```diff
@@ -363,7 +388,12 @@ private void channelConnectionFutureHandler(@Nonnull ChannelFuture future,
+                if (sslAuthenticationFailure.get()) {
+                    retryPolicy.nextWait();
+                } else {
+                    retryPolicy.reset();
```
Consider turning ExponentialBackoffRetry to an immutable class and assigning a new instance rather than reseting.

annym commented on runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java:
```diff
@@ -312,6 +313,9 @@
         int systemDownHandlerTriggerLimit = 20;
 
+        @Default
+        Duration maxRetryThreshold = Duration.ofMinutes(5);
```
This is already in the CorfuRuntimeParameters, same as RPC timeout.

annym commented:
@Maithem I have addressed all comments, please review. Exponential backoff for reconnections is now performed regardless of the cause that triggered the reconnection and resetting the retry connection time calculator is only performed on connection success.

Maithem commented on runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java:
```diff
@@ -186,7 +187,7 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
     public void exceptionCaught(ChannelHandlerContext ctx,
         java.lang.Throwable cause) throws Exception {
-        log.error("exceptionCaught: Exception {} caught.", cause.getClass().getSimpleName(), cause);
+        log.error("exceptionCaught: Exception {} caught. Cause: {}", cause.getClass().getSimpleName(), cause);
```
Does this make https://github.com/CorfuDB/CorfuDB/pull/1572/files#diff-74369a0020e71f6d710749351f0ba98eR475 redundant ?

annym replied:
Not really right? the other code is on server logs.

zalokhan commented on runtime/src/main/java/org/corfudb/util/retry/ExponentialBackoffRetry.java:
```diff
@@ -60,8 +63,13 @@ public ExponentialBackoffRetry(IRetryable runFunction) {
+
     @Override
-    public void nextWait() {
+    public long nextWait() {
```
Can we return Duration?
